src/shared/Core/Trace2Exception.cs

```diff
@@ -0,0 +1,67 @@
+using System;
+using System.ComponentModel;
+using System.IO;
+using GitCredentialManager.Authentication.OAuth;
+using GitCredentialManager.Interop;
+
+namespace GitCredentialManager;
+
+public class Trace2Exception : Exception
+{
+    public Trace2Exception(ITrace2 trace2, string message) : base(message)
+    {
+        trace2.WriteError(message);
+    }
+
+    public Trace2Exception(ITrace2 trace2, string message, string messageFormat) : base(message)
+    {
+        trace2.WriteError(message, messageFormat);
+    }
+}
+
+public class Trace2InvalidOperationException : InvalidOperationException
+{
+    public Trace2InvalidOperationException(ITrace2 trace2, string message) : base(message)
+    {
+        trace2.WriteError(message);
+    }
+}
+
+public class Trace2OAuth2Exception : OAuth2Exception
+{
+    public Trace2OAuth2Exception(ITrace2 trace2, string message) : base(message)
+    {
+        trace2.WriteError(message);
+    }
+}
+
+public class Trace2InteropException : InteropException
+{
+    public Trace2InteropException(ITrace2 trace2, string message, int errorCode) : base(message, errorCode)
+    {
+        trace2.WriteError($"message: {message} error code: {errorCode}");
+    }
+
+    public Trace2InteropException(ITrace2 trace2, string message, Win32Exception ex) : base(message, ex)
+    {
+        trace2.WriteError(message);
+    }
+}
+
+public class Trace2GitException : GitException
+{
+    public Trace2GitException(ITrace2 trace2, string message, string gitMessage, int errorCode) :
```

Comment:
I stumbled a bit here because this method takes two strings and I was expecting the second to be an instance of a `messageFormat` but it is a different creature.  I realize that that comes from the base class, but still it caused me to double check the caller and callee a few times.

Not sure there is an action item here, just making a note here.

Reply:
maybe move the error code earlier in the calling sequence, so we would see (effectively)
the current (and later if we have to add the fmt version):

```
Trace2GitException(iTrace2, string,         int, string gitMessage)
Trace2GitException(iTrace2, string, string, int, string gitMessage)
```